persistence.js

```diff
@@ -627,8 +657,94 @@ class RedisPersistence extends CachedPersistence {
     this._db.eval(luaScript, 4, [topic, groupTopic, clientTopic, SHAREDTOPICS], cb)
   }
 
+  fillSharedCache (cb) {
+    const that = this
+    that._db.smembers(SHAREDTOPICS, function (err, sharedTopics) {
+      if (err) {
+        return cb(err)
+      }
+      if (!sharedTopics) {
+        return cb(null)
+      }
+
+      that.sharedCachedQlobber = new Qlobber(qlobberOpts)
+      that.cachedSharedTopics = new Set(sharedTopics)
+      const pipeline = that._db.pipeline()
+      sharedTopics.forEach((topicFromResult) => {
+        that.sharedCachedQlobber.add(topicFromResult, topicFromResult)
+        pipeline.smembers(topicFromResult)
+      })
+
+      // Execute pipeline
+      const allClientTopics = []
+      pipeline.exec((err, replies) => {
+        if (err) {
+          cb(err)
+        } else {
+          // Map replies to topics
+          that.cachedSharedTopicsToGroups = new Map()
+          replies.forEach((reply, index) => {
+            const topic = sharedTopics[index]
+            that.cachedSharedTopicsToGroups.set(topic, new Set(reply[1]))
+            if (reply[1] && reply[1].length > 0) {
+              reply[1].forEach((clientTopic) => {
+                allClientTopics.push(clientTopic)
+              })
+            }
+          })
+          const pipelineForGroups = that._db.pipeline()
+          allClientTopics.forEach((item) => {
+            pipelineForGroups.smembers(item)
+          })
+          pipelineForGroups.exec((err, groupsReplies) => {
+            if (err) {
+              cb(err)
+            } else {
+              // Map replies to topics
+              that.cachedSharedGroupsClientTopics = new Map()
+              groupsReplies.forEach((reply, index) => {
+                that.cachedSharedGroupsClientTopics.set(allClientTopics[index], new Set(reply[1]))
+              })
+
+              that.lastCacheRefreshTs = (new Date() / 1000)
+              cb(null)
+            }
+          })
+        }
+      })
+    })
+  }
+
+  getSharedTopicsFromCache (topic, cb) {
+    const resultTopics = []
+    const matches = this.sharedCachedQlobber.match(topic)
+    for (const match of matches) {
+      if (this.cachedSharedTopics.has(match)) {
+        const groups = this.cachedSharedTopicsToGroups.get(match)
+        for (const group of groups) {
+          const clientTopics = Array.from(this.cachedSharedGroupsClientTopics.get(group))
+          const randomTopic = clientTopics[Math.floor(Math.random() * clientTopics.length)]
+          resultTopics.push(randomTopic + topic)
+        }
+      }
+    }
+    cb(null, resultTopics)
+  }
+
   getSharedTopics (topic, cb) {
-    const luaScript = `
+    if (this.sharedCacheRefreshIntervalSec) {
+      if (this.lastCacheRefreshTs + this.sharedCacheRefreshIntervalSec < (new Date() / 1000)) {
```

Comment:
This is more obvious for me and does the same:
```suggestion
      if (this.lastCacheRefreshTs + this.sharedCacheRefreshIntervalSec < (Date.now() / 1000)) {
```